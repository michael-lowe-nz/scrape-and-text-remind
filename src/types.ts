export type Contacts = {
  Teams: Array<Team>;
};

export type Team = {
  Name: string;
  Players: Array<{ Name: string; Number: string }>;
  ScheduleURL: string;
};

export type Game = {
  dateString: string;
  time: string;
  court: string;
  teamAgainst: string;
};
