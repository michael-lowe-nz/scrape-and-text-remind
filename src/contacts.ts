const examplePhoneNumber = "+18005550100";

export default {
  Test: {
    Teams: [
      {
        Name: "Los Angeles Lakers",
        ScheduleURL:
          "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX",
        Players: [
          {
            Name: "Shaquille ONeal",
            Number: process.env.TEST_PHONE_NUMBER || examplePhoneNumber,
          },
        ],
      },
    ],
  },
};
