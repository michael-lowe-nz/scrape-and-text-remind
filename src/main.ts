import { existsSync, readFileSync } from "fs";
import { App, Aspects, Stage, StageProps } from "aws-cdk-lib";
import { AwsSolutionsChecks, HIPAASecurityChecks } from "cdk-nag";
import { Construct } from "constructs";
import { load } from "js-yaml";
import ContactData from "./contacts";
import { LeagueLobsterTextReminder } from "./stacks/leagueLobsterTextReminders";
import { OIDCSetup } from "./stacks/oidcSetup";
import { TestStack } from "./stacks/testStack";
import { Contacts } from "./types";

const testEnv = {
  account: "653221278763",
  region: "ap-southeast-2",
};

const devEnv = {
  account: "476203294330",
  region: "ap-southeast-2",
};

const prodEnv = {
  account: "365979456435",
  region: "ap-southeast-2",
};

const app = new App();

new OIDCSetup(app, "oidc-setup");

let localContacts: any;
let prodContacts: any;

if (existsSync("./src/contacts.yml")) {
  const localContactsYml = readFileSync("./src/contacts.yml", "utf-8");
  localContacts = load(localContactsYml);
} else {
  localContacts = ContactData.Test;
}

if (process.env.CONTACTS_YML) {
  prodContacts = load(process.env.CONTACTS_YML);
} else {
  prodContacts = ContactData.Test;
}

interface StageTemplateProps extends StageProps {
  contacts: Contacts;
  environmentName: string;
  runOnSchedule: boolean;
  runTestsOnDeploy: boolean;
}

class TextRemindersStage extends Stage {
  constructor(scope: Construct, id: string, props: StageTemplateProps) {
    super(scope, id, props);

    const stack = new LeagueLobsterTextReminder(this, "TextReminders", {
      Contacts: props.contacts,
      EnvironmentName: props.environmentName,
      RunOnSchedule: props.runOnSchedule,
    });

    if (props.runTestsOnDeploy) {
      new TestStack(this, "TestStack", {
        RemindLambdaFunction: stack.testFunction,
      });
    }

    Aspects.of(stack).add(new AwsSolutionsChecks({ verbose: true }));
    Aspects.of(app).add(new HIPAASecurityChecks({ verbose: true }));
  }
}

new TextRemindersStage(app, "dev-stage", {
  env: devEnv,
  contacts: localContacts,
  environmentName: "Dev",
  runOnSchedule: false,
  runTestsOnDeploy: false,
});

new TextRemindersStage(app, "test-stage", {
  env: testEnv,
  contacts: ContactData.Test,
  environmentName: "Test",
  runOnSchedule: false,
  runTestsOnDeploy: true,
});

new TextRemindersStage(app, "prod-stage", {
  env: prodEnv,
  contacts: prodContacts,
  environmentName: "Production",
  runOnSchedule: true,
  runTestsOnDeploy: false,
});

app.synth();
