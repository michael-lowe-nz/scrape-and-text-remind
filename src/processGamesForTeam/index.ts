import { PublishCommand, SNSClient } from "@aws-sdk/client-sns";
import { Handler } from "aws-lambda";
import axios from "axios";
// import { getRandomEmoji } from "./lib/getRandomEmoji";
import moment from "moment";
import {
  extractGamesFromHTML,
  getNextGameAfterDate,
  getGameInfoFromGame,
} from "../lib/gamesFunctions";
import { Game } from "../types";

const url = process.env.SCHEDULE_URL;

export const handler: Handler = async () => {
  const client = new SNSClient({ region: process.env.AWS_REGION });

  const teamName = process.env.TEAM_NAME;

  /**
   * @todo should be an environment variable
   */

  console.log("URL", url);

  const requestResponse: any = await axios.get(url || "");
  const games: Array<Game> = extractGamesFromHTML(requestResponse.data);
  console.log("GAMES:");
  console.log(games);
  const nextGame: Game | null = getNextGameAfterDate(games, moment());

  if (!nextGame) {
    const Message = `Ok, looks like we can't find a game for ${teamName} the coming week. You might want to check what is going on! üëÄ`;
    const params = {
      Message,
      TopicArn: process.env.SNS_ADMIN_TOPIC_ARN,
      DefaultSMSType: "Promotional",
    };
    const response = await client.send(new PublishCommand(params));
    console.log("No game message sent");
    console.log(response);
    return {
      statusCode: 200,
      body: {
        message: "No game coming up this week",
      },
    };
  }

  const gameInfo: string = getGameInfoFromGame(nextGame);
  const Message = `üèÄ ${teamName} Reminders üèÄ\nHey guys! Game is on ${gameInfo}`;
  const params = {
    Message,
    TopicArn: process.env.SNS_TOPIC_ARN,
    DefaultSMSType: "Promotional",
  };
  console.log("Game Message Sent:");
  console.log(Message);
  const response = await client.send(new PublishCommand(params));
  return {
    statusCode: 200,
    body: { snsResponse: response, message: Message },
  };
};
