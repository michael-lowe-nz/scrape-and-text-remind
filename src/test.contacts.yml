---
Teams:
  - Name: Gen-X
    ScheduleURL: "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX"
    Players:
      - Name: Michael
        Number: "+18005550100"
        IsAdmin: true
  - Name: "Craig's Team"
    ScheduleURL: "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX"
    Players:
      - Name: "Craig"
        Number: "+18005550100"
