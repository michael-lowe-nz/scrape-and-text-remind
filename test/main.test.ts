import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
  LeagueLobsterTextReminder,
  LeagueLobsterTextReminderProps,
} from "../src/stacks/leagueLobsterTextReminders";
import { Contacts } from "../src/types";

const testContacts: Contacts = {
  Teams: [
    {
      Name: "Test Team",
      ScheduleURL:
        "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX",
      Players: [
        {
          Name: "Test Player",
          Number: "+1",
          IsAdmin: true,
        },
        {
          Name: "Another Player",
          Number: "+2",
        },
      ],
    },
    {
      Name: "Test Team 2",
      ScheduleURL:
        "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX",
      Players: [
        {
          Name: "Test Player",
          Number: "+3",
        },
        {
          Name: "Another Player",
          Number: "+4",
        },
      ],
    },
  ],
};

const TestProps: LeagueLobsterTextReminderProps = {
  Contacts: testContacts,
};

test("Snapshot", () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test", TestProps);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

test("Test that the stack builds an SNS topic", () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test", TestProps);

  const template = Template.fromStack(stack);
  expect(template.hasResource("AWS::SNS::Topic", {}));
});

test("Test that the Text reminder stack builds a lambda function to remind each team", () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test", TestProps);
  const template = Template.fromStack(stack);
  expect(template.resourceCountIs("AWS::Lambda::Function", 2));
});
