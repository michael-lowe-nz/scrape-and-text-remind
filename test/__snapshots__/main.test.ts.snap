// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlertTestTeam2Function6D2517FA": {
      "DependsOn": [
        "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850",
        "AlertTestTeam2FunctionServiceRoleEA0E27EA",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "92bc77498ebaf10aef21e5a285682ab7bae1eed978a063d49f5fea134dfa2609.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_URL": "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX",
            "SNS_TOPIC_ARN": {
              "Ref": "TestTeam2Alerts2BE53530",
            },
            "TZ": "Pacific/Auckland",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AlertTestTeam2FunctionServiceRoleEA0E27EA",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "TestTeam2Alerts2BE53530",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850",
        "Roles": [
          {
            "Ref": "AlertTestTeam2FunctionServiceRoleEA0E27EA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AlertTestTeam2FunctionServiceRoleEA0E27EA": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlertTestTeamFunction3257EF79": {
      "DependsOn": [
        "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B",
        "AlertTestTeamFunctionServiceRole496082A0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "92bc77498ebaf10aef21e5a285682ab7bae1eed978a063d49f5fea134dfa2609.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_URL": "https://websites.mygameday.app/team_info.cgi?c=0-2854-0-633233-27116561&a=SFIX",
            "SNS_TOPIC_ARN": {
              "Ref": "TestTeamAlerts58363266",
            },
            "TZ": "Pacific/Auckland",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AlertTestTeamFunctionServiceRole496082A0",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlertTestTeamFunctionServiceRole496082A0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "TestTeamAlerts58363266",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B",
        "Roles": [
          {
            "Ref": "AlertTestTeamFunctionServiceRole496082A0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTeam2Alerts2BE53530": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":kms:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":alias/aws/sns",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
    "TestTeam2Alerts362D23E2E": {
      "Properties": {
        "Endpoint": "+3",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeam2Alerts2BE53530",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeam2Alerts4320BD087": {
      "Properties": {
        "Endpoint": "+4",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeam2Alerts2BE53530",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeam2RuleAllowEventRuletestAlertTestTeam2Function3AFE01234D49C8F2": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AlertTestTeam2Function6D2517FA",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TestTeam2RuleB81240E8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestTeam2RuleB81240E8": {
      "Properties": {
        "ScheduleExpression": "cron(30 8 ? * SUN *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "AlertTestTeam2Function6D2517FA",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TestTeamAlerts17FF09393": {
      "Properties": {
        "Endpoint": "+1",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeamAlerts58363266",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeamAlerts2E7EDE869": {
      "Properties": {
        "Endpoint": "+2",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeamAlerts58363266",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeamAlerts58363266": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":kms:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":alias/aws/sns",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
    "TestTeamRule8F21550D": {
      "Properties": {
        "ScheduleExpression": "cron(30 8 ? * SUN *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "AlertTestTeamFunction3257EF79",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TestTeamRuleAllowEventRuletestAlertTestTeamFunction959B320A6E3ED324": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AlertTestTeamFunction3257EF79",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "TestTeamRule8F21550D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
